epan/dissectors/packet-ippusb.c

```diff
@@ -21,70 +21,421 @@
+    }
+    else if (!pinfo->fd->visited) {
+
+         //first time this packet is ever seen
+
+        gboolean save_fragmented = pinfo->fragmented;
+        pinfo->fragmented = TRUE;
+
+        ti = proto_tree_add_item(tree, proto_ippusb, tvb, offset, -1, 0);
+        ippusb_tree = proto_item_add_subtree(ti, ett_ippusb);
+
+
+        if(is_http_header(first_linelen, first_line)){
+
+            if (ippusb_last_pdu >= 0) {
+                ippusb_last_pdu = -1;
+            }
+
+            new_msp = pdu_store(pinfo, ippusbd->multisegment_pdus);
+            new_msp->running_size = reported_length;
+
+            fragment_add_check(&ippusb_reassembly_table, tvb, offset, pinfo, new_msp->first_frame,
+                                        GUINT_TO_POINTER(new_msp->first_frame), 0, reported_length, TRUE);
+            ippusb_last_pdu = pinfo->num;
+        }
+        else{
+
+            previous_msp = (struct ippusb_multisegment_pdu *)wmem_tree_lookup32_le(ippusbd->multisegment_pdus, ippusb_last_pdu);
+
+            if(previous_msp){
+                previous_msp->nxtpdu = pinfo->num;
+                new_msp = pdu_store(pinfo, ippusbd->multisegment_pdus);
+                new_msp->first_frame = previous_msp->first_frame;
+                new_msp->running_size = previous_msp->running_size + reported_length;
+
+                if(last != 0x03) {
+                    fragment_add_check(&ippusb_reassembly_table, tvb, offset, pinfo, new_msp->first_frame,
+                                        GUINT_TO_POINTER(new_msp->first_frame), previous_msp->running_size, reported_length, TRUE);
+                    ippusb_last_pdu = pinfo->num;
+                }
+                else {
+                    new_msp->finished = TRUE;
+
+                    fragment_head * head = fragment_add_check(&ippusb_reassembly_table, tvb, offset, pinfo, new_msp->first_frame,
+                                                        GUINT_TO_POINTER(new_msp->first_frame), previous_msp->running_size, reported_length, FALSE);
+                    tvbuff_t *processedTvb = process_reassembled_data(tvb, offset, pinfo, "Reassembled IPPUSB", head, &ippusb_frag_items, NULL, tree);
+
+                    pinfo->can_desegment = 0;
+
+                    ret = dissector_try_uint_new(ippusb_dissector_table, HTTP, processedTvb, pinfo, tree, TRUE, data);
+                    col_append_fstr(pinfo->cinfo, COL_INFO, "  Reassembled Data");
+                    ippusb_last_pdu = -1;
+
+                }
+            }
+        }
+
+        pinfo->fragmented = save_fragmented;
+    }
+    else {
+
+        //not the first time this is seen
+
+        gboolean save_fragmented = pinfo->fragmented;
+        pinfo->fragmented = TRUE;
+        current_msp = (struct ippusb_multisegment_pdu *)wmem_tree_lookup32_le(ippusbd->multisegment_pdus, pinfo->num);
+
+        if (current_msp && current_msp->nxtpdu == 0 && !current_msp->finished) {
```

Comment:
>{ [](start = 79, length = 1)

For the if/else blocks since some of the logic is non-obvious, include a comment at the top of the block to describe why it was selected (something like // Not finished processing current multi-segment PDU and this is the last Protocol Data Unit)